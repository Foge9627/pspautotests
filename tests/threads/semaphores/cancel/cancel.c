#include "../sub_shared.h"

SETUP_SCHED_TEST;

#define CANCEL_TEST(title, sema, count) { \
	int result = sceKernelCancelSema(sema, count, NULL); \
	if (result == 0) { \
		printf("%s: OK\n", title); \
	} else { \
		printf("%s: Failed (%X)\n", title, result); \
	} \
	PRINT_SEMAPHORE(sema); \
}

#define CANCEL_TEST_WITH_WAIT(title, sema, count) { \
	int waitThreads = 99; \
	int result = sceKernelCancelSema(sema, count, &waitThreads); \
	if (result == 0) { \
		printf("%s: OK (%d waiting)\n", title, waitThreads); \
	} else { \
		printf("%s: Failed (%X, %d waiting)\n", title, result, waitThreads); \
	} \
	PRINT_SEMAPHORE(sema); \
}

int main(int argc, char **argv) {
	SceUID sema = sceKernelCreateSema("cancel", 0, 0, 1, NULL);

	CANCEL_TEST("Normal", sema, 1);
	CANCEL_TEST("Greater than max", sema, 3);
	CANCEL_TEST("Zero", sema, 0);
	CANCEL_TEST("Negative -3", sema, -3);
	CANCEL_TEST("Negative -1", sema, -1);

	CANCEL_TEST_WITH_WAIT("Normal", sema, 1);
	CANCEL_TEST_WITH_WAIT("Greater than max", sema, 3);
	CANCEL_TEST_WITH_WAIT("Zero", sema, 0);
	CANCEL_TEST_WITH_WAIT("Negative -3", sema, -3);
	CANCEL_TEST_WITH_WAIT("Negative -1", sema, -1);

	sceKernelDeleteSema(sema);

	TWO_STEP_SCHED_TEST(0, 1,
		CANCEL_TEST_WITH_WAIT("To 0", sema1, 0);
	,
		CANCEL_TEST_WITH_WAIT("To 1", sema1, 1);
	);

	CANCEL_TEST("NULL", 0, 0);
	CANCEL_TEST("Invalid", 0xDEADBEEF, 0);
	CANCEL_TEST("Deleted", sema, 0);

	return 0;
}