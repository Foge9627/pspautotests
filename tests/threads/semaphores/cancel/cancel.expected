Normal: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Greater than max: Failed (800201BD)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Zero: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -3: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -1: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Normal: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Greater than max: Failed (800201BD, 99 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Zero: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -3: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -1: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
ABTo 0: OK (0 waiting)
Sema: OK (size=56,name='schedTest1',attr=0,init=0,cur=0,max=1,wait=1)
CDTo 1: OK (1 waiting)
Sema: OK (size=56,name='schedTest1',attr=0,init=0,cur=1,max=1,wait=0)
E
NULL: Failed (80020199)
Sema: Failed (00000000)
Invalid: Failed (80020199)
Sema: Invalid (80020199)
Deleted: Failed (80020199)
Sema: Invalid (80020199)
